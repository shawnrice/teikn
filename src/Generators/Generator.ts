import { EOL } from 'os';

import { version } from '../../package.json';
import { Plugin } from '../Plugins';
import { Token } from '../Token';
import { matches } from '../utils';

export type GeneratorOptions = {
  /**
   * The extension for the file
   *
   * (this is usually set when the Generator is extended)
   */
  ext: string;
  /**
   * The basename for the file
   *
   * default: `tokens`
   */
  filename?: string;
};

export type RequiredGeneratorOptions = {
  ext: string;
};

export type RequiredGeneratorOptionNames = keyof RequiredGeneratorOptions;

export abstract class Generator<Opts extends GeneratorOptions = GeneratorOptions> {
  options: Opts;

  constructor(opts: Opts) {
    this.options = opts;
    this.validateOptions();

    this.signature = this.signature.bind(this);
  }

  signature(): string {
    return `Teikn v${version}`;
  }

  validateOptions(): void {
    const required: RequiredGeneratorOptions = { ext: 'string' };

    const errors: string[] = [];

    for (const key in required) {
      const type = typeof this.options[key as RequiredGeneratorOptionNames];
      const expected = required[key as RequiredGeneratorOptionNames];

      if (type !== expected) {
        errors.push(`Error: received option ${key} of type ${type}; expected ${expected}.`);
      }
    }

    if (errors.length) {
      throw new Error(errors.join(EOL));
    }
  }

  get file(): string {
    const { ext, filename = 'tokens' } = this.options;

    return [filename, ext].join('.');
  }

  header(): string | null {
    return null;
  }

  footer(): string | null {
    return null;
  }

  protected prepareTokens(tokens: Token[], plugins: Plugin[]): Token[] {
    return tokens.map(token =>
      plugins.reduce((acc, plugin) => {
        if (!matches(plugin.tokenType, token.type)) {
          return acc;
        }

        if (!matches(plugin.outputType, this.options.ext)) {
          return acc;
        }

        return plugin.toJSON(acc);
      }, token),
    );
  }

  abstract generateToken(_: Token): any;

  abstract combinator(_: Token[]): string;

  generate(tokens: Token[], plugins: Plugin[] = []): string {
    return [this.header(), this.combinator(this.prepareTokens(tokens, plugins)), this.footer()]
      .filter(Boolean)
      .join(EOL)
      .trim();
  }
}

export default Generator;
