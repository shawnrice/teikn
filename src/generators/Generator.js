const { EOL } = require('os');

class Generator {
  constructor(opts) {
    this.options = opts;
    this.validateOptions();
  }

  validateOptions() {
    const required = { ext: 'string' };

    const errors = [];

    Object.keys(required).forEach(name => {
      const type = typeof this.options[name];
      const expected = required[name];

      if (type !== expected) {
        errors.push(`Error: received option ${name} of type ${type}; expected ${expected}.`);
      }
    });

    if (errors.length) {
      throw new Error(errors.join(EOL));
    }
  }

  get file() {
    const { ext, filename = 'tokens' } = this.options;

    return [filename, ext].join('.');
  }

  header() {
    return null;
  }

  footer() {
    return null;
  }

  generateToken(token) {
    throw new Error('You need to extend the Generator class');
  }

  combinator() {
    throw new Error('You need to extend the Generator class');
  }

  generate(tokens) {
    return [this.header(), this.combinator(tokens), this.footer()]
      .filter(Boolean)
      .join(EOL)
      .trim();
  }
}

module.exports = Generator;
