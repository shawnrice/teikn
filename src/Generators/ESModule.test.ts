import { default as tokenSet1 } from '../fixtures/tokenSet1';
import Generator from './ESModule';

describe('ESModule tests', () => {
  test('it generates tokens as an esmodule', () => {
    expect(new Generator({ dateFn: () => 'null' }).generate(tokenSet1)).toMatchSnapshot();
  });
});
