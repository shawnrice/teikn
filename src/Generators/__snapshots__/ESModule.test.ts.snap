// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESModule tests it generates tokens as an esmodule 1`] = `
"/**
 * Teikn v1.0.0-alpha.5
 * Generated null
 *
 * This file is generated and should be commited to source control
 *
 */
export const tokens = {
  /**
   *  Use for body fonts
   *  Type: font-family
   */
  body: '\\"Roboto Condensed\\", Arial, sans',
  /**
   *  the primary branding color
   *  Type: color
   */
  primary: 'aliceblue',
};


export default tokens;"
`;
